.github/workflows/run_tests.yml

```diff
@@ -0,0 +1,106 @@
+# SPDX-FileCopyrightText: TNG Technology Consulting GmbH <https://www.tngtech.com>
+#
+# SPDX-License-Identifier: Apache-2.0
+
+name: Install and Run tests
+
+on:
+  push:
+    branches:
+      - main
+  pull_request:
+  workflow_dispatch:
+
+jobs:
+  linter:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ ubuntu-latest, macos-latest ]
+        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11" ]
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: cache poetry install
+        uses: actions/cache@v3
+        with:
+          path: ~/.local
+          key: poetry-${{matrix.os}}-1.4.0-0
+
+      - name: Install poetry
+        uses: snok/install-poetry@v1
+        with:
+          version: 1.4.0
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+      - name: cache deps
+        id: cache-deps
+        uses: actions/cache@v3
+        with:
+          path: .venv
+          key: pydeps-${{ matrix.os}}-${{matrix.python-version }}-${{ hashFiles('**/poetry.lock') }}
+
+      - name: Install package dependencies
+        run: poetry install --no-interaction --no-root
+        if: steps.cache--deps.outputs.cache-hit != 'true'
+
+      - run: poetry install --no-interaction
+
+      - name: Run isort
+        run: poetry run isort src/ tests/
+
+      - name: Run black
+        run: poetry run black src/ tests/
+
+      - name: Run flake8
+        run: poetry run flake8 src/ tests/
+
+      - name: Run mypy
+        run: poetry run python -m mypy src/ tests/
+
+  test:
+    runs-on: ${{ matrix.os }}
+    strategy:
+      matrix:
+        os: [ ubuntu-latest, macos-latest ]
+        python-version: [ "3.7", "3.8", "3.9", "3.10", "3.11" ]
+
+    steps:
+      - uses: actions/checkout@v3
+      - name: Set up Python ${{ matrix.python-version }}
+        uses: actions/setup-python@v4
+        with:
+          python-version: ${{ matrix.python-version }}
+
+      - name: cache poetry install
+        uses: actions/cache@v3
+        with:
+          path: ~/.local
+          key: poetry-${{matrix.os}}-1.4.0-0
+
+      - name: Install poetry
+        uses: snok/install-poetry@v1
+        with:
+          version: 1.4.0
+          virtualenvs-create: true
+          virtualenvs-in-project: true
+      - name: cache deps
+        id: cache-deps
+        uses: actions/cache@v3
+        with:
+          path: .venv
+          key: pydeps-${{ matrix.os}}-${{matrix.python-version }}-${{ hashFiles('**/poetry.lock') }}
+
+      - name: Install package dependencies
+        run: poetry install --no-interaction --no-root
+        if: steps.cache--deps.outputs.cache-hit != 'true'
+
+      - run: poetry install --no-interaction
```

Comment:
Optional: As this part is duplicated, I'm asking myself if there is some way in Github actions to extract this somehow to avoid dupliaction? If it would be too much work to refactor it might be fine to keep it as it is right now.

Reply:
I tried to extract these steps and came across this [blogpost](https://cardinalby.github.io/blog/post/github-actions/dry-reusing-code-in-github-actions/) which might be useful. Unfortunately, I wasn't able to get this to work properly. As I think this is reasonable and would be an improvement, I will open an issue for this, so that we can address this later.

Reply:
Great :+1: